### PR TITLE
i#1356: Break down instrumentation pass

### PR DESCRIPTION
**Pull request by johnfxgalea**
Breaks down the code intended for copying cb information to local temporary storage into smaller helper functions. The main instrumentation pass of a basic block is also moved to a separate helper function.

This is a preliminary PR to help with the integration of drbbdup and drmgr.

Issue: #1356

### CODE REVIEW

**Comment by johnfxgalea** [comment]
@derekbruening  I referenced issue #1356 because it relates to the local copying of cb information - which this PR refactors. I don't think I need to create a new issue for this PR but LMK if you want me to.

File: ext/drmgr/drmgr.c [hunk 1]
```diff
@@ -992,18 +973,79 @@ drmgr_bb_event(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
...
+     * expensive. Instead, we create a scoped table later on that only maps the cb lists
+     * of those opcodes required by this specific bb.
+     */
+    dr_rwlock_read_unlock(bb_cb_lock);
+}
+
+static void
+drmgr_bb_event_delete_local_cb_info(void *drcontext, IN local_cb_info_t *local_info)
+{
+
```
Comment (derekbruening) [inline ext/drmgr/drmgr.c:1011]: nit: blank line

File: ext/drmgr/drmgr.c [hunk 2]
```diff
@@ -992,18 +973,79 @@ drmgr_bb_event(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
...
+{
+    dr_emit_flags_t res;
+    local_cb_info_t local_info;
+    void **pair_data = NULL, **quartet_data = NULL;
+    per_thread_t *pt = (per_thread_t *)drmgr_get_tls_field(drcontext, our_tls_idx);
+
+    drmgr_bb_event_set_local_cb_info(drcontext, &local_info);
+
+    /* We need per-thread user_data */
+    if (pair_count > 0)
```
Comment (derekbruening) [inline ext/drmgr/drmgr.c:1032]: This needs to be merged with the race fix -- so this needs to be moved inside the _set_ function with the lock and added to the local struct or sthg right?

Comment (johnfxgalea) [reply ext/drmgr/drmgr.c]: Yep, done

File: ext/drmgr/drmgr.c (hunk 1 above)
Comment (johnfxgalea) [reply ext/drmgr/drmgr.c]: Done.

File: ext/drmgr/drmgr.c (hunk 2 above)
Comment (johnfxgalea) [reply ext/drmgr/drmgr.c]: I tried to break down the PRs for easier reviews. Hope it didn't yield confusion instead.

File: ext/drmgr/drmgr.c [hunk 3]
```diff
@@ -1029,20 +1033,22 @@ drmgr_bb_event(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
...
 
-    if (pair_count > 0)
-        dr_thread_free(drcontext, pair_data, sizeof(void *) * pair_count);
-    if (quartet_count > 0)
-        dr_thread_free(drcontext, quartet_data, sizeof(void *) * quartet_count);
+    if (local_info.pair_count > 0)
+        dr_thread_free(drcontext, pair_data, sizeof(void *) * local_info.pair_count);
+    if (local_info.quartet_count > 0)
+        dr_thread_free(drcontext, quartet_data,
+                       sizeof(void *) * local_info.quartet_count);
```
Comment (derekbruening) [inline ext/drmgr/drmgr.c:1051]: style: {} around multi-line

File: ext/drmgr/drmgr.c (hunk 2 above)
Comment (derekbruening) [reply ext/drmgr/drmgr.c]: No, small is good.

**Comment by johnfxgalea** [comment]
Great, thank you for the reviews!

File: ext/drmgr/drmgr.c (hunk 3 above)
Comment (johnfxgalea) [reply ext/drmgr/drmgr.c:1052]: Done.

**Pull request merged**